dimod/generators/qap.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2024 D-Wave Inc.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+
+from itertools import product
+
+import numpy as np
+
+from dimod.binary_quadratic_model import BinaryQuadraticModel
+from dimod.constrained import ConstrainedQuadraticModel
+
+__all__ = ['quadratic_assignment']
+
+def quadratic_assignment(distance_matrix: np.typing.ArrayLike,
+                         flow_matrix: np.typing.ArrayLike,
+                         ) -> ConstrainedQuadraticModel:
+    """Generates a constrained quadratic model encoding a quadratic assignment problem.
+
+    Given distance and flow matrices, generates a :class:`~dimod.ConstrainedQuadraticModel` 
+    for the corresponding quadratic assignment problem. 
+
+    Args:
+        distances: numpy array containing the distance between location i and location j
+
+        flows: numpy array containing the flow between facility i and facility j
+
+    Returns:
+        The constrained quadratic model encoding the quadratic assignment problem. Variables are
+        denoted as ``x_{i}_{j}`` where ``x_{i}_{j} == 1`` means that facility ``i`` is
+        placed in location ``j``.
+
+    """
+
+    if distance_matrix.shape != flow_matrix.shape:
+        raise ValueError("'distance_matrix' and 'flow_matrix' must have the same shape")
+
+    num_locations = distance_matrix.shape[0]
+
+    model = ConstrainedQuadraticModel()
+    obj = BinaryQuadraticModel(vartype='BINARY')
+    x = {(i, j): obj.add_variable(f'x_{i}_{j}') for i in range(num_locations)
+         for j in range(num_locations)}
+
+    for i, j, k, l in product(range(num_locations), repeat=4):
+        if (i,j) != (k,l):
+            obj.set_quadratic(x[(i,j)], x[(k,l)], flow_matrix[i][k]*distance_matrix[j][l]
+                              + flow_matrix[k][i]*distance_matrix[j][l])
+
+    model.set_objective(obj)
+
+    for i in range(num_locations):
+        constraint = [(x[(i,j)], 1) for j in range(num_locations)] + [(-1,)]
+        model.add_constraint(constraint, sense='==')
+
+    for j in range(num_locations):
```

Comment:
We could/should add one of these as a [discrete constraint](https://docs.ocean.dwavesys.com/en/stable/docs_dimod/reference/generated/dimod.ConstrainedQuadraticModel.add_discrete.html).

Reply:
Thank you! All suggestions are applied.